File: js/factorymethod/Factory.java
package factorymethod;

public abstract class Factory {
    public Item create(String name) {
        boolean bCreatable = this.isCreatable(name);
        if (bCreatable) {
            Item item = this.createItem(name);
            postprocessItem(name);
            return item;
        }
        return null;
    }

    protected abstract void postprocessItem(String name);
    protected abstract Item createItem(String name);
    protected abstract boolean isCreatable(String name);
}

File: ts/src/factory_method/Factory.ts
import Item from "./Item";

export type ItemName = "sword" | "shield" | "bow";

export default abstract class Factory {
  protected abstract isCreatable(name: ItemName): boolean;
  protected abstract createItem(name: ItemName): Item;
  protected abstract postProcessItem(name: ItemName): void;

  create(name: ItemName): Item {
    const isCreatable = this.isCreatable(name);
    if (isCreatable) {
      const item = this.createItem(name);
      this.postProcessItem(name);
      return item;
    }
    return null;
  }
}
